### PR TITLE
Migrate TickBitmap tests from HardHat to Foundry 

### PR DESCRIPTION
**Pull request by gitcoindev**
## Related Issue
This pull request migrates TickBitmap tests from HardHat to Foundry (35 test cases), a part of https://github.com/Uniswap/v4-core/issues/315 .

## Description of changes
- migrated 35 TickBitmap tests to Foundry
- removed HardHat implementation after migration

```
Running 35 tests for test/foundry-tests/TickBitmap.t.sol:TickBitmapTestTest
[PASS] test_flipTick_flipsOnlyTheSpecifiedTick() (gas: 37394)
[PASS] test_flipTick_gasCostOfFlippingATickThatResultsInDeletingAWord() (gas: 19391)
[PASS] test_flipTick_gasCostOfFlippingFirstTickInWordToInitialized() (gas: 28230)
[PASS] test_flipTick_gasCostOfFlippingSecondTickInWordToInitialized() (gas: 29492)
[PASS] test_flipTick_revertsOnlyItself() (gas: 83555)
[PASS] test_isInitialized_isFalseAtFirst() (gas: 8323)
[PASS] test_isInitialized_isFlippedBackByFlapTick() (gas: 20466)
[PASS] test_isInitialized_isFlippedByFlapTick() (gas: 29793)
[PASS] test_isInitialized_isNotChangedByAnotherFlipToADifferentTick() (gas: 29549)
[PASS] test_isInitialized_isNotChangedByAnotherFlipToADifferentTickOnAnotherWord() (gas: 33421)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_doesNotExceedBoundary() (gas: 65229)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_gasCostForEntireWord() (gas: 65074)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_gasCostJustBelowBoundary() (gas: 63073)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_gasCostOnBoundary() (gas: 65074)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_returnsTheNextInitializedTickFromTheNextWord() (gas: 86853)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_returnsTheNextWordsInitializedTickIfOnTheRightBoundary() (gas: 65242)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_returnsTheNextWordsInitializedTickIfOnTheRightBoundary2() (gas: 63663)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_returnsTheTickDirectlyToTheRight() (gas: 63610)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_returnsTheTickDirectlyToTheRight2() (gas: 63700)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_returnsTickToRightIfAtInitializedTick() (gas: 63641)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_returnsTickToRightIfAtInitializedTick2() (gas: 63681)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_skipsEntireWord() (gas: 65240)
[PASS] test_nextInitializedTickWithinOneWord_lteFalse_skipsHalfWord() (gas: 65208)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_atTheWordBoundary() (gas: 65226)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_boundaryIsInitialized() (gas: 86703)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_entireEmptyWord() (gas: 65203)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_gasCostForEntireWord() (gas: 65081)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_gasCostJustBelowBoundary() (gas: 63379)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_gasCostOnBoundary() (gas: 65070)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_halfwayThroughEmptyWord() (gas: 65214)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_returnsSameTickIfInitialized() (gas: 63512)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_returnsTickDirectlyToTheLeftOfInputTickIfNotInitialized() (gas: 63503)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_willNotExceedTheWordBoundary() (gas: 65216)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_wordBoundary() (gas: 65256)
[PASS] test_nextInitializedTickWithinOneWord_lteTrue_wordBoundaryLess1nextInitializedTickInNextWord() (gas: 63506)
Test result: ok. 35 passed; 0 failed; finished in 7.48ms
```

### CODE REVIEW

**Comment by gitcoindev** [comment]
> * function test_isInitialized_isFlippedByFlapTick()**
> * function test_isInitialized_isFlippedBackByFlapTick()**
> * I think the name should be Flip instead of Flap.

good catch!

**Comment by gitcoindev** [comment]
@ewilz I fixed typos and refactored tests to use forge-gassnapshot, PR is ready for the review.

**Comment by gitcoindev** [comment]
The PR needs the same changes as https://github.com/Uniswap/v4-core/pull/317 , in progress...

**Comment by gitcoindev** [comment]
Hi @ewilz this PR is ready for another review round

**Review by nagrarohit — APPROVED** [review]
LGTM. But i missed the following comments :- 
// #isInitialized
// #flipTick
// #nextInitializedTickWithinOneWord

**Comment by ewilz** [comment]
hardhat snapshots need to be deleted

File: test/foundry-tests/TickBitmap.t.sol
Comment (ewilz) [inline test/foundry-tests/TickBitmap.t.sol:108]: maybe stick this function towards the bottom of the test file in keeping with organization structure of other forge tests

**Comment by ewilz** [comment]
Can delete TickBitmapTest.sol and rename foundry test `TickBitmapTest` hence getting rid of the extra "Test" on there

**Comment by ewilz** [comment]
test migration looks good, not sure if you've ever written forge fuzz tests, with the recently committed FullMath conversion for instance, we took the echidna fuzz tests and converted them to forge fuzz tests, need to convert `checkNextInitializedTickWithinOneWordInvariants` Happy to merge without that though and just do it later..

**Comment by gitcoindev** [comment]
Hi @ewilz thank you for the review, I already pushed changes. I can have a look at fuzzing later this week after merging.

**Comment by ewilz** [comment]
woops sorry didn't think I had write access to your branch

**Review by ewilz — APPROVED** [review]

**Comment by gitcoindev** [comment]
Yes I enabled write access, thanks for merging!

**Pull request merged**
